### PR TITLE
Sandbox/Locking prototypes

### PR DESCRIPTION
**Pull request by Lusus**
This pull request contains a couple of prototypes which may be handy to add to tbot_contrib. They most certainly commit crimes against humanity in their current form.

The first patch contains a U-Boot Sandbox board configuration which can be useful to get started with Tbot, without actually requiring real hardware.

The second patch contains 3 different machine locking implementations, and uses the Sandbox as testbed. The first implementation is in Python/bash. The second uses procmail's lockfile utility, which served as "good example" for the first implementation. The third implementation uses the dotlockfile utility.

In the Sandbox, a default locking pool size of 2 has been configured, which means two Sandbox instances can be started before an exception will be thrown. The timeout value specifies for how long these locks are valid, but is commented out by default. With short lock timeouts and/or interactive_uboot sessions, the locks my expire without the instance noticing, which causes problems. The requirements around this may need to be reconsidered.

### CODE REVIEW

File: tbot_contrib/locking.py [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+from tbot.machine import machine, linux
+
+
+class MachineLock(machine.PreConnectInitializer):
+    """
+    Machine locking implementation based on Python and bash.
+    """
+    @property
+    def lock_name(self) -> str:
+        return self.__class__.__name__
```
Comment (Rahix) [inline tbot_contrib/locking.py:14]: Use `self.name` here.

File: tbot_contrib/locking.py [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+    @property
+    def lock_pool_max(self) -> int:
+        return 1
+
+    @property
+    def lock_sleep(self) -> int:
+        return 1
+
+    lock_no = -1
+    lock_file = ""
```
Comment (Rahix) [inline tbot_contrib/locking.py:29]: If I understand correctly, those are actually instance attributes, not class attributes.  As such, please don't define them on the class.

File: tbot_contrib/locking.py [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+    """
+    Machine locking implementation based on Python and bash.
+    """
+    @property
+    def lock_name(self) -> str:
+        return self.__class__.__name__
+
+    @property
+    def lock_expiry(self) -> int:
+        return -1
```
Comment (Rahix) [inline tbot_contrib/locking.py:18]: In python, we generally use `None` to mark the absence of a value instead of e.g. `-1`.  You can then check for it with

```python
if variable is not None:
    ...

# or

if variable is None:
    ...
```

File: tbot_contrib/locking.py [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+
+    @property
+    def lock_pool_max(self) -> int:
+        return 1
```
Comment (Rahix) [inline tbot_contrib/locking.py:22]: For "const" values, there is no need to use `@property`.  Just use a class attribute assignment:

```python
lock_pool_max: int = 1
```

File: tbot_contrib/locking.py [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+
+    lock_no = -1
+    lock_file = ""
+
+    @contextlib.contextmanager 
+    def _init_pre_connect(self) -> typing.Iterator:
+        with tbot.ctx.request(tbot.role.LabHost) as lh:
+            result = lh.exec("mktemp", "lock.XXX")
+            if result[0] == 0:
+                tempfile = result[1].strip()
```
Comment (Rahix) [inline tbot_contrib/locking.py:36]: You probably want to make this a `Path`, then the `RedirStdout` below will also work:

```python
tempfile = linux.Path(lh, result[1].strip())
```

File: tbot_contrib/locking.py [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+                            if (int(result[1]) != 0) and (int(result[1]) < int(time.time())):
+                                lh.exec("rm", self.lock_file) # Delete stale lock
+                                print(f"Lock {self.lock_file} expired, deleted.")
+                                repeat = True; # The stale lock may be available next time round.
+                if self.lock_no != -1: # If we got a lock, check that is still ours
+                    time.sleep(self.lock_sleep) # Allow possible multiple expired lock deletes to settle
+                    result = lh.exec("stat", "-c%i", self.lock_file)
+                    if result[0] == 0: # lock still exists
+                        if result[1] != lh.exec("stat", "-c%i", tempfile)[1]:
+                            repeat = True; # but inodes don't match, try again
```
Comment (Rahix) [inline tbot_contrib/locking.py:70]: Decompose directly instead of then having to deal with a tuple and indexing:

```python
retcode, output = lh.exec("stat", "-c%i", self.lock_file)
if retcode == 0: # lock still exists
    if output != lh.exec("stat", "-c%i", tempfile)[1]:
        repeat = True; # but inodes don't match, try again
```

File: tbot_contrib/locking.py [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+    def lock_pool_max(self) -> int:
+        return 1
+
+    lock_no = -1
+    lock_file = ""
+
+    @contextlib.contextmanager 
+    def _init_pre_connect(self) -> typing.Iterator:
+        with tbot.ctx.request(tbot.role.LabHost) as lh:
+            if tbot.tc.shell.check_for_tool(lh, "lockfile") == False:
```
Comment (Rahix) [inline tbot_contrib/locking.py:108]:
```suggestion
            if not tbot.tc.shell.check_for_tool(lh, "lockfile"):
```

File: tbot_contrib/locking.py [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+                raise Exception("Could not create tempfile")
+            repeat = True
+            while (repeat == True):
+                repeat = False
+                lock_no = -1
+                for count in range(self.lock_pool_max): # Iterate through locks
+                    self.lock_file = f"{self.lock_name}{count}"
+                    print(f"Try locking on {count} using {self.lock_file}...")
+                    result = lh.exec("ln", tempfile, self.lock_file) # Is atomic 
+                    if result[0] == 0: # We got a lock, we're done
```
Comment (Rahix) [inline tbot_contrib/locking.py:53]: Use `LinuxMachine.test()` if you just care about success of a command:

```suggestion
                    if lh.test("ln", tempfile, self.lock_file): # We got a lock, we're done
```

**Review by Rahix — CHANGES_REQUESTED** [review]
Hmm, I would drop the lock pool from being a part of the locking implementation.  It blows up the complexity quite a bit and makes it really hard to follow the locking algorithm.

Instead, I think the `MachineLock` should only concern itself with acquiring and releasing a _single_ lock.  One can then build a separate mechanism around the `MachineLock` which attempts locking for a list of names, and use this to implement the pool as a separate thing.  The way it is right now, the pooling and the locking are interleaved which makes it hard to differentiate which part does what.

I've left a few comments inline to hint you at a few things where there are more pythonic replacements :)

**Comment by Rahix** [comment]
When you have time, please rebase on the current `master` to pull in the CI changes that have happened.

**Comment by Lusus** [comment]
Okay, I tried my best to apply all of your comments, and I rebased on master. Hopefully I didn't mess anything up too badly. ;-)

I reordered the implementations as follows:

The first is what I think/hope is the best of the bunch, it is in Python/shell, supports expiry, but also checks whether the "owner" is still around before deleting stale locks.
The second is similar, but does not check the owner. This does what the spec asks.
The third uses lockfile
The fourth uses dotlockfile

I kept the pooling in the contextmanager/_init_pre_connect() functions, because I think hooking it in another way may get tricky. These function then calls _locker() (all with the same signature) which does the actual locking in the various way.

File: tbot_contrib/locking.py [hunk 9]
```diff
@@ -0,0 +1,302 @@
...
+                            )
+            if (
+                success
+            ):  # Because things could race, test whether we still have the lock...
+                time.sleep(self.lock_sleep)  # Allow things to settle
+                # Check whether the lockfile still exists and get its inode.
+                retval, result = labhost.exec("stat", "-c%i", lockfile)
+                if retval == 0:  # lock still exists
+                    # Get inode of the tempfile, to see if it matches lockfile's
+                    if result != labhost.exec("stat", "-c%i", tempfile)[1]:
```
Comment (Rahix) [inline tbot_contrib/locking.py:86]: I think it would be much more readable to use the following construction:

```python
try:
    lockstat = lockfile.stat()
except OsError:
    repeat = True
else:
    if lockstat.st_ino != tempfile.stat().st_ino:
        repeat = True 
```

File: tbot_contrib/locking.py [hunk 10]
```diff
@@ -0,0 +1,302 @@
...
+            labhost.exec("echo", linux.Raw("$$"), linux.Raw(">>"), tempfile)
+            return tempfile
+        else:
+            raise Exception("Could not create tempfile")
+
+    def _locker(self, labhost, lockfile) -> bool:
+        tempfile = self._make_temp(labhost)
+        success = False
+        repeat = True
+        while repeat:  # do-while
```
Comment (Rahix) [inline tbot_contrib/locking.py:55]: I'm not sure I understand what this loop is for.  We only attempt acquiring the lock once, right?  Or am I missing something here?

**Review by Rahix — CHANGES_REQUESTED** [review]
Hi,

there's a lot to talk about; I'm not sure how to organize it neatly.  I hope my reply isn't too much all over the place ^^

#### 1. CI Errors & `pre-commit`
CI is still complaining about a few things.  You can check in the [Checks](https://github.com/Rahix/tbot/pull/45/checks) tab of this PR what errors it encountered.  You can also run the linters/tests locally by using the [`pre-commit`](https://pre-commit.com/) tool :)

#### 2. High-level architecture
Right now you've implemented the locking directly as part of the "board machine".  This is problematic because it means an unlock will happen immediately when the board is turned off instead of later, when the entire session is torn down.

My suggestion for this would be to move the locking into the lab-host, by means of a `LockManager`.  This is where all the dirty bits lie, the initializer for the board just goes and asks the selected lab-host for a specific lock.  As a rough sketch:

```python
class LockManagerBase(abc.ABC):
    """
    Defines the "interface" that each locking implementation needs to
    implement.
    """
    @abc.abstractmethod
    def request_machine_lock(self, name: str) -> bool:
        raise tbot.error.AbstractMethodError()


class SomeLockManager(LockManagerBase, machine.PostShellInitializer):
    """
    A concrete locking implementation.

    This class should be inherited by the lab-host.
    """
    def request_machine_lock(self, name: str) -> bool:
        # Try locking `name` here.
        if we_got_the_lock:
            self._active_locks.append(name)
            return True
        else:
            return False

    def _release_machine_lock(self, name: str) -> None:
        # Release the lock `name`.
        ...

    @contextlib.contextmanager
    def _init_post_shell(self) -> Iterator:
        self._active_locks = []
        try:
            yield None
        finally:
            # Locks are only released when the lab-host exits.
            for lock in self._active_locks():
                self._release_machine_lock(lock)


class MachineLock(machine.PreConnectInitializer):
    """
    This is the initializer that is inherited by the board machine.  It just
    calls the lab-host's locking implementation.
    """
    @contextlib.contextmanager
    def _init_pre_connect(self) -> typing.Iterator:
         with tbot.ctx.request(tbot.role.LabHost) as lh:
            if not isinstance(lh, LockManagerBase):
                raise Exception("selected lab-host is not a lock manager")
            if not lh.request_machine_lock(self.lock_name):
                raise Exception("could not acquire the lock")

            yield None
```

I hope this makes sense?

#### 3. Separation of the pooling implementation
I think with the above, this becomes super easy.  Pooling is just another initializer for the board machine, which is separate from the `MachineLock` from above.  For example:

```python
class PooledMachineLock(machine.PreConnectInitializer):
    """
    A 'replacement' for `MachineLock` which acquires a lock from a pool.
    """
    @contextlib.contextmanager
    def _init_pre_connect(self) -> typing.Iterator:
         with tbot.ctx.request(tbot.role.LabHost) as lh:
            if not isinstance(lh, LockManagerBase):
                raise Exception("selected lab-host is not a lock manager")

            for lock_name in candidates:
                if lh.request_machine_lock(lock_name):
                    # We got the lock, everything is alright
                    break
            else:
                # None of the locks were free
                raise Exception("could not acquire the lock")

            tbot.log.message(f"Acquired lock {lock_name!r}!")
            yield None
```

---

For continuing, I would suggest let's stick with getting one implementation working completely first and re-adding the others again later.  Otherwise you'll have a lot more work keeping them all in sync as things change.  Would that be okay for you?

**Comment by Lusus** [comment]
Okay, I guess I will leave my comments here then.

About the CI, yes, I see the CI is a bit stricter than Pylint... Will fix

About the high level architecture. The main issue why it is currently in the board machine, is because the board would actually need to translate the lock number, to an actual machine in the pool, where the assumption was, we have a pool of machines, represented by the board machine definition. Are we ever intending to use a pool of locks for dissimilar machines? The lock should not get teared down when the machine gets turned off, as we are using the _init_pre_connect() context, and the lock should remain for the life of the lab session instead. And I know there is work in progress for making lab sessions live longer. ;-) But I will definitely try to do it as you requested (if I can figure it out).

The only reason there are/were multiple implementations of the locking is to make sure I don't code myself into a corner, and to try to make issues obvious. I hope that I spotted most of the gotchas, so I will stick with only the first implementation from now on. :-)

File: tbot_contrib/locking.py [hunk 11]
```diff
@@ -0,0 +1,302 @@
...
+            if self.lock_expiry < self.lock_sleep:  # if lock does not expire
+                labhost.exec("echo", "0", linux.RedirStdout(tempfile))
+            else:  # else write the Unix time timestamp of expiry
+                labhost.exec(
+                    "echo",
+                    str(int(time.time()) + self.lock_expiry),
+                    linux.RedirStdout(tempfile),
+                )
+            # Append the PID of this context's shell
+            labhost.exec("echo", linux.Raw("$$"), linux.Raw(">>"), tempfile)
```
[the remark this answers is not in the document]
Comment (Lusus) [reply tbot_contrib/locking.py:46]: Thanks, I will do that. I did not know I could create files on the remote host in this way.

File: tbot_contrib/locking.py [hunk 12]
```diff
@@ -0,0 +1,302 @@
...
+                    break
+            if self.lock_no is None:
+                raise Exception("Could not get free lock")
+            try:
+                yield None
+            finally:
+                labhost.exec("rm", "-f", self.lock_file)
+
+
+class MachineLock2(machine.PreConnectInitializer):
```
[the remark this answers is not in the document]
Comment (Lusus) [reply tbot_contrib/locking.py:115]: Look, the only reason the second implementation exists, is because it does what I understood the spec to request. The problem is, if a lock "expires" but the previous "owner" is still using it, the lock would get re-used without the "owner" knowing about it, which creates issues which are beautifully illustrated by multiple Sandboxes; the 3rd implementation using the "lockfile" utility suffers from exactly the same issue. The idea for storing the PID comes from the 4th implementation, which in turn does not expire.

So the goal is to only keep "first" implementation. The other implementations remain only as references for now, partly to show that multiple implementations can co-exist. They are also duplicated so that modifying one doesn't break the others.

File: tbot_contrib/locking.py [hunk 13]
```diff
@@ -0,0 +1,302 @@
...
+
+        labhost.exec("rm", tempfile)  # Delete the temp file
+        return success
+
+    @contextlib.contextmanager
+    def _init_pre_connect(self) -> typing.Iterator:
+        with tbot.ctx.request(tbot.role.LabHost) as labhost:
+            self.lock_no = None
+            for count in range(self.lock_pool_max):  # iterate through locks
+                self.lock_file = linux.Path(labhost, f"{self.lock_name}{count}")
```
[the remark this answers is not in the document]
Comment (Lusus) [reply tbot_contrib/locking.py:102]: Absolutely! I will use the directory you suggested. It was just easier to deal with while playing around with the prototypes.

File: tbot_contrib/locking.py [hunk 14]
```diff
@@ -0,0 +1,302 @@
...
+                            print(f"Expired lock {lockfile} PID still appears valid!")
+                        else:
+                            labhost.exec("rm", lockfile)  # Delete stale lock
+                            print(f"Lock {lockfile} expired, deleted.")
+                            repeat = (
+                                True  # The stale lock may be available next time round.
+                            )
+            if (
+                success
+            ):  # Because things could race, test whether we still have the lock...
```
[the remark this answers is not in the document]
Comment (Lusus) [reply tbot_contrib/locking.py:80]: The locking is indeed atomic, but the deleting of a stale lock isn't unfortunately. I'll explain below...

File: tbot_contrib/locking.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (Rahix) [reply tbot_contrib/locking.py:46]: [`Path.write_text()`](https://tbot.tools/modules/machine_linux.html#tbot.machine.linux.Path.write_text) essentially amounts to `tee filename >/dev/null` and sending the contents to it:

https://github.com/Rahix/tbot/blob/067e9998b8abe611a06cf8c3e4ce704dc98d3205/tbot/machine/linux/path.py#L218-L233

File: tbot_contrib/locking.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (Rahix) [reply tbot_contrib/locking.py:115]: Ah, okay, I was under the impression you want to keep both.  No problem then!

File: tbot_contrib/locking.py [hunk 15]
```diff
@@ -0,0 +1,302 @@
...
+                time.sleep(self.lock_sleep)  # Allow things to settle
```
[the remark this answers is not in the document]
Comment (Lusus) [reply tbot_contrib/locking.py:81]: Yes, you are right, the sleeping avoids the fact that we can't atomically evaluate and delete a stale lock.

We read the lock to determine whether it is stale, check for the PID, and only then delete it. It is possible that multiple lock seekers try to delete the same lock. In the worst case, a lock can be deleted, then atomically locked again by someone else, and then deleted again by still someone else thinking it is still the same old lock.

Sleeping before checking whether we still have the lock was the only workaround I could think of.

File: tbot_contrib/locking.py [hunk 16]
```diff
@@ -0,0 +1,302 @@
...
+                if retval == 0:
+                    results = result.split()
+                    # Check if the lock can expire, and whether it has expired.
+                    if (int(results[0]) != 0) and (int(results[0]) < int(time.time())):
+                        # Check whether the lock PID is still in use by a bash instance
+                        retval, result = labhost.exec("ps", "-ocomm=", results[1])
+                        if (retval == 0) and (result.strip() == "bash"):
+                            print(f"Expired lock {lockfile} PID still appears valid!")
+                        else:
+                            labhost.exec("rm", lockfile)  # Delete stale lock
```
[the remark this answers is not in the document]
Comment (Lusus) [reply tbot_contrib/locking.py:73]: Exactly. Please, no locks locking locks! ;-)

File: tbot_contrib/locking.py (hunk 9 above)
Comment (Lusus) [reply tbot_contrib/locking.py:86]: Thanks for that suggestion, I did not know of this handy utility.

File: tbot_contrib/locking.py (hunk 10 above)
Comment (Lusus) [reply tbot_contrib/locking.py:55]: The loop is there for when the lock was stale and deleted, so that we can try locking it again. Or for when we got the lock, but it someone else deleted it since we took it (as explained above).

**Comment by Rahix** [comment]
> About the CI, yes, I see the CI is a bit stricter than Pylint... Will fix

Regarding the errors about `lock_no` and `lock_file` in particular:  Those should be instance attributes (that means attributes on an object of the type inheriting this class) while you're defining class attributes here (attributes on the class directly).  You probably want to not define them on the class and only add them dynamically to the instance later on, during initialization.

> The main issue why it is currently in the board machine, is because the board would actually need to translate the lock number, to an actual machine in the pool, where the assumption was, we have a pool of machines, represented by the board machine definition. Are we ever intending to use a pool of locks for dissimilar machines?

Maybe I should clarify, the sketch I posted above does support this.  The `PooledMachineLock` would be inherited by the board-machine and can make the name of the acquired lock available for later steps during instantiation:

```python
class PooledMachineLock(machine.PreConnectInitializer):
    @contextlib.contextmanager
    def _init_pre_connect(self) -> typing.Iterator:
         with tbot.ctx.request(tbot.role.LabHost) as lh:
            for lock_name in candidates:
                if lh.request_machine_lock(lock_name):
                    break
            else:
                raise Exception("could not acquire the lock")

            # Save name of acquired lock:
            self.active_lock = lock_name

            yield None
```

And then later you could, for example, use this when connecting:

```python
class MyBoard(connector.ConsoleConnector, PooledMachineLock, board.Board):
    def connect(self, mach):
        return mach.open_channel("picocom", f"/dev/tty{self.active_lock}")
```

> The lock should not get teared down when the machine gets turned off, as we are using the `_init_pre_connect()` context, and the lock should remain for the life of the lab session instead.

`_init_pre_connect` is still only a part of the board machines lifecycle, even if it is the first thing to run and the last to exit (think of it like the outermost shell of an onion maybe).  So when the board machine is deinitialized, the `_init_pre_connect` context will exit as well.  The lab-host lifetime is not at play here.

> The only reason there are/were multiple implementations of the locking is to make sure I don't code myself into a corner, and to try to make issues obvious. I hope that I spotted most of the gotchas, so I will stick with only the first implementation from now on. :-)

Ah, got it, that wasn't clear to me, sorry!

**Comment by Lusus** [comment]
Okay, so I rebuilt everything in the form that you suggested, and I managed to get everything by pre-commit and friends. Whether it is correct is a second matter. I had to go back to self.lock_no = -1 to make it happen though. I am certain that there are multiple style issues outstanding.

To make testing the Sandbox more feasible, I added a locallab.py machine config and an args file (in the first commit). To ease the testing of the locking using the Sandbox, I added the lock-sandbox.py and lock-locallab.py machine configs and an args file (in the second commit).

The idea is, that by selecting the corresponding args file, the uboot_build and interactive_uboot testcases can then build/run the Sandbox locally. By running interactive_uboot in multiple terminals the locking can also be demonstrated.

**Review by Rahix — CHANGES_REQUESTED** [review]
Hey!

It is really taking shape now :) I've left a few comments in places where there are more idiomatic solutions, I hope they help you get a bit better grasp of what's available to you in Python.

As we discussed out of band, I think the locking algorithm complexity can be drastically reduced by not trying to deal with the racey expiry check and instead adding synchronization to it using (e.g.) `flock(1)`.

File: tbot_contrib/locking.py [hunk 17]
```diff
@@ -0,0 +1,216 @@
...
+
+
+class LockManagerBase(abc.ABC):
+    """
+    Defines the "interface" that each locking implementation needs to
+    implement.
+    """
+
+    @abc.abstractmethod
+    def request_machine_lock(self, properties: dict) -> bool:
```
[the remark this answers is not in the document]
Comment (Lusus) [reply tbot_contrib/locking.py:18]: The reason I used a dictionary as parameter, is that I did not want the derived class to dictate the signature of the parent, as I could imagine other LockManagers wanting other parameters, and these not being relevant to them. Is there perhaps a better way as to not have to specify all the detailed parameters in the parent than a dictionary? If this is not an issue from your view, I will happily change things as you described.

[the remark this answers is not in the document]
Comment (Rahix) [reply tbot_contrib/locking.py:18]: I think it is not a good idea to allow too much flexibility in the parameters, because it will create pitfalls in terms of compatibility.  Ideally the `MachineLock` should not care about the lock-manager in use at all so that you can reuse a board config with any lab-config no matter which lock-manager it uses.

Thus I'd say it is actually a _feature_ to restrict the parameters which can be passed.

However, you're completely right that e.g. `checkpid` is not really a generic parameter, it really only applies to the one `LockManager` implementation which we have now.  `expiry` on the other hand is generic, I'd say; it is a setting which each lock-manager should provide.

I think the best solution would be to just move `checkpid` into the `LockManager` entirely and not allow `MachineLock` to customize it.  I.e. the `LockManager` gets a new `lock_checkpid` configuration option to configure whether pid-checking should be enabled on this lab-host and anything `checkpid` related is dropped from the `MachineLock`.  What do you think?

---

For completeness:  If you want to allow customization of the signature, you could design the "interface class" like this:

```python
from typing import Optional, Any

class LockManagerBase(abc.ABC):
    @abc.abstractmethod
    def request_machine_lock(
        self,
        name: str,
        *,
        expiry: Optional[int] = None,
        **kwargs: Any,
    ) -> bool:
        ...
```

This `kwargs` "catches" all additional keyword arguments and (in a function with a body) would be a dictionary with their values.

Though as written above, I don't think this is the right approach here.

[the remark this answers is not in the document]
Comment (Lusus) [reply tbot_contrib/locking.py:18]: I am actually happy with moving the checkpid to LockManager, as it leaves less room for confusion. I guess if someone later requires additional arguments for another LockManager, we can do what you suggested here, but not definitely not for checkpid! Thanks for this advice.

**Comment by Lusus** [comment]
Okay, so I tried my best to apply all your previous fixes, and then I built a prototype using flock as a master lock to serialise the checking and removal of stale locks as FlockManager.

Testing has not yet been as extensive as for LockManager, and there are still details outstanding like docstrings, etc

**Comment by Lusus** [comment]
Sorry, forgot to rebase my branch on master...

**Review by Rahix — CHANGES_REQUESTED** [review]
Thanks!  I left a few more comments about the new implementation, but overall it looks fine.  I'd say once those are resolved we'll merge the locking and then do further work with the implementation being in-tree.

One thing I'd like to keep out for now is the sandbox config:  I'll have to spend some more time looking at that and especially taking a closer look at this change:

```diff
@@ -76,7 +76,7 @@ def close(self) -> None:
             raise channel.ChannelClosedException()
 
         sid = os.getsid(self.p.pid)
-        self.p.terminate()
+        self.p.kill()
         os.close(self.pty_master)
         self.p.wait()
```

So I'd say we should restrict this PR to just the locking module and then add the sandbox stuff later on if that's okay with you.

File: tbot_contrib/locking.py [hunk 18]
```diff
@@ -0,0 +1,299 @@
...
+            self._active_locks.add(lockfile)
+        return success
+
+    # The problem here is, if someone else reclaimed and reused a lock we held,
+    # we may now remove it here...
+    def release_machine_lock(self, name: str) -> None:
+        lockfile = self.lock_dir / name
+        if lockfile in self._active_locks:
+            self._active_locks.remove(lockfile)
+            self.exec("rm", "-f", lockfile)
```
[the remark this answers is not in the document]
Comment (Rahix) [reply tbot_contrib/locking.py:286]: Looking at this more, I think this line should also be dropped:

```python
if lockfile in self._active_locks:
```

If someone calls `release_machine_lock()` for a lock that is not held, this indicates bad downstream code and thus we should fail with noise instead of silently ignoring it.

**Comment by Rahix** [comment]
Note: Once you're done please also mark this PR as "ready for review" as it is no longer only a draft :)

**Comment by Lusus** [comment]
I applied the changes as you had requested, and tried to apply the principles consistently everywhere. I also tried to add docstrings etc. everywhere. I removed the Sandbox components from this branch as you requested; they are still available for convenient testing at: https://github.com/Lusus/tbot/tree/lusus-sandbox

**Review by Rahix — APPROVED** [review]
> I applied the changes as you had requested, and tried to apply the principles consistently everywhere. I also tried to add docstrings etc. everywhere.

Looks great, thanks for working on this :)  I'll give it a spin against a "real" project tomorrow, just as a sanity check.  Unless I stumble over something we missed so far, I'll merge this PR afterwards.

> I removed the Sandbox components from this branch as you requested; they are still available for convenient testing at: https://github.com/Lusus/tbot/tree/lusus-sandbox

That's handy, thanks!
